introduction: replace out-dated revault.pdf with introduction.md (#113)
revault/practical-revault: introduction.md

> ---

Author: > Actually, again re the diagram i find it confusing it's in term of transactions and not outputs. Do you find it better conveys the flow that the PDF's diagram (which i find to be perfectly and immediately clear)?

I think they are both clear. Grouping into transactions and using descriptors abstraction matches the text better. Also this one shows which transactions are pre-signed (will make that and clock clearer with white background). What is confusing about it?

> ---

Reviewer: > What is confusing about it?

What *i find* confusing is that it's in terms of transactions instead of outputs. What if a transaction pays to 3 deposit outputs? Also, is it clear the Cancel, UnEmer, and Spend all spend the same output?

Again, i'm not saying it's a valid reason to change it. Just expressing purely subjective feedback.

> ---

Author: Yeah good point. I think details of multiple outputs or batching are low-level and best left for transactions.md not the intro. But I can see how it could be unclear that the cancel, unemer and spend consume the same output. I'll extend the caption and make sure it's clear.

header-includes:
  - \hypersetup{colorlinks=true,
            allbordercolors={0 0 0},
            pdfborderstyle={/S/U/W 1}}

> ---

Author: ah yeah sorry... I used it for converting into nice PDF with pandoc

# Revault: A Multi-Party Bitcoin Vault Architecture

The secure storage and transaction of funds is a big challenge to Bitcoin users and especially to companies managing a substantial amount of bitcoins. Traditional controls of expenses are hard to set-up with bitcoin due to its features: censorship-resistance, transaction irreversibility and psuedonymous addresses.

Revault is a self-managed custody solution for multi-party holders (such as a company or their third party custodian) with strong theft mitigation from external parties and internal fund managers. Revault has several novel features; 

- On-chain tiered access control, where stakeholders have primary control and an ability to safely delegate funds to managers
- Ability to cancel managers' withdrawals from a vault (automatically based on pre-configured policy, or manually)
- Ability to restrict managers' spending behaviours based on arbitrary policies
- Physical coercion mitigation using a trapdoor mechanism to move all funds to a preset `emergency vault`
- Ability to scale-up security investment simply by deploying more internal or external watchtowers.

> Revault can be tailor-fit to many organisational structures with custom multi-signature thresholds, time-lock periods, and feature sets. Users can set policies in-line with business logic. Users can start with a simple deployment and scale infrastructure as their security budget increases. 

Author: I think (unvault policy) cancel & emergency are a significant part of the deployment security. 

Arguably, with well-defined ceremonies (initial and for updating participants) it's also feasible to scale-up the stakeholders and/or managers sets (e.g. once a year)

## Process and Revault Transaction Types

![Revault Transaction Graph. ](Revault-Tx-Diagram.png)

Funds enter custody through Deposit transactions (TXs) as unspent-transaction-outputs (UTXOs) locked to the `deposit output descriptor`, an N-of-N multi-signature among stakeholders. After receiving a deposit, the stakeholders exchange signatures for a set of 4 transactions which are _not broadcast_. These "pre-signed" transactions are used by stakeholders to control the flow of funds and enable the delegation and emergency deterrent features of Revault. As seen in the diagram, they include; Emergency, Unvault, Cancel and Unvault Emergency TXs. 

The Emergency TX is the first to be signed. It would spend a deposit and pay to the `emergency output descriptor` which is a bitcoin script that is harder to unlock than `deposit output descriptor`. Keeping the `emergency output descriptor` private among stakeholders strengthens the emergency deterrent feature as attackers don't know how they would compromise the emergency vault.

> Then, stakeholders exchange signatures for the Cancel and Unvault Emergency TXs:

Reviewer: It's how the practical protocol currently is specified. But [it's going to change](https://github.com/revault/practical-revault/issues/109) and in any case this document shouldn't be influenced by the details of the protocol.

So, yeah, it should be merged together.

> Then, stakeholders exchange signatures for the Cancel and Unvault Emergency TXs:

Author: > in any case this document shouldn't be influenced by the details of the protocol.

I find this confusing. What is "the protocol" for you? (I think you mean messages?) 

In my opinion the intro should be consistent with the rest of practical-revault, so if messages are passed in a certain order (as in messages.md) that should be the same in the intro. Happy to merge these given the upcoming change tho

- The Unvault Emergency TX would spend the Unvault TX to pay to the `emergency output descriptor`.
- The Cancel TX would spend the Unvault TX to pay back to the `deposit output descriptor`.

This way, funds are protected _before_ delegation to fund managers. Finally, the Unvault TX is signed. It spends the Deposit TX and pays to either:

> - the managers' K-of-M multi-signature + J cosigning servers + time-lock of X blocks, or 

Reviewer: I think a setup with a number of cosigs different than the number of stakeholders (or say one operated by an independent organization instead of the stakeholders themselves) makes sense, but it's not actually specified nor part of Revault. Like watchtowers with only Emergency or Cancel. Therefore i don't think we should mention having a different number of cosigs here.

> - the managers' K-of-M multi-signature + J cosigning servers + time-lock of X blocks, or 

Author: Yeah. I'm adding "J = 0 if deploying without co-signing servers, J = N if deploying with co-signing servers"

- to the stakeholders' N-of-N multi-signature without delay. 

> Managers can attempt to use the Spend TX to pay external wallets, but are forced to wait for the Unvault time-lock to expire before their transaction finalizes. During that delay period, a Cancel or Unvault Emergency TX can be broadcasted if the manager is breaching pre-defined policy. 

Reviewer: > Broadcast is the correct past afaik?

Both work (i looked up recently :)).

>  I would remove "if the manager is breaching pre-defined policy", as it does not apply to Emer nor manual Cancel.

Agree.

Please check the complete specification of [transactions](transactions.md) for more details.

## Infrastructure and Feature Choice

Users can choose features to suit their needs. Different features require different infrastructure to operate. 

First, there are base requirements common to all Revault deployments. Each stakeholder and manager controls a wallet and a signing device. An untrusted Coordinator server is used to route messages and thus enable wallets to update and synchronise their state.  

> - **Wallet**: an open-source software running with a bitcoind back-end used to keep track of the users' co-owned coins, transaction signatures, and to connect with a signing device and infrastructure servers.

Reviewer: > I wouldn't even describe a wallet as software, but anyways

How would you describe it? :no_mouth: 
A wallet is a software connecting to the Bitcoin network and keeping track of coins ownership among all existing UTxOs (and maybe your own tx history, too).

>  This is nitpicking, please ignore if irrelevant.

Yeah, no opinion. I see how it's nice to have as it makes for a concrete instance for people to model how would Revault be actually used without reading up all the specs.

> - **Wallet**: an open-source software running with a bitcoind back-end used to keep track of the users' co-owned coins, transaction signatures, and to connect with a signing device and infrastructure servers.

Author: I think if I define it as "Revault Wallet" and add "track history" that would address the issues?

- **Signing Device**: a hardware module used to store private keys, public keys and wallet descriptors and to handle signing operations. As user verification is an important part of the Revault security model, such a device must allow the user to check the message integrity before signing it. 

- **Coordinator server**: An always-online server used for efficient message routing among participants and other servers. Its primary use is for exchanging signatures and Spend TXs. 

With these in place, stakeholders are able to delegate bitcoin to fund managers and manually broadcast Cancel, Emergency and Unvault-Emergency TXs. Fund managers will also be able to manually broadcast Cancel TXs to recover from mistakes. 

Next, users who want additional features will require additional infrastructure components. Each stakeholder may control one or more watchtowers and/ or a co-signing server.

> - **Watchtower**: An always-online server that monitors the bitcoin network and chain, as well as polls the coordinator for Spend TXs to perform its responsibilities (automated unvault policy enforcement, automated spend policy enforcement, emergency deterrent). 

Reviewer 2: And the WT only needs the Spend in case of a Cosig deployment OR weird extra sig policy like Darosior's ones.

> - **Co-signing Server**: An always-online server that acts as a 'stupidly simple' anti-replay signing oracle: it will accept to sign any Spend TX, *but only once*. This is necessary to stop the managers from changing the Spend TX destination after the Unvault TX’s time-lock has expired. 

Reviewer 2: Will sign only each Unvault UTXO once. Spend is irrelevant as such.

> - **Co-signing Server**: An always-online server that acts as a 'stupidly simple' anti-replay signing oracle: it will accept to sign any Spend TX, *but only once*. This is necessary to stop the managers from changing the Spend TX destination after the Unvault TX’s time-lock has expired. 

Reviewer: > from changing the Spend TX destination after the Unvault TX’s time-lock has expired.

Huh, no. After timelock expiration if the Unvault was rogue but not revaulted, it's game over.

```suggestion
- **Co-signing Server**: An always-online server that acts as a 'stupidly simple' anti-replay signing oracle: it will accept to sign any Spend TX, *but only once*. This is necessary to stop the managers from malleating the Spend transaction after the Unvault transaction(s) were broadcast. 
```

Automatic unvault policy enforcement; requires a Watchtower. By default the Watchtower will broadcast a Cancel TX any time it detects an Unvault TX unless it adheres to a specific pre-configured unvault policy.

> Automatic spend policy enforcement; requires a Watchtower and Co-signing Server. By default the Watchtower will broadcast a Cancel TX any time it detects the Spend TX's parent Unvault TX unless it is made aware of the Spend TX in advance and that Spend TX adheres to a pre-configured spend policy. The co-signing server is a means to stop the managers from changing the Spend TX after the Unvault TX's time-lock has expired. 

Reviewer 2: to spend the unvault with another transaction, more precisely.

> Automatic Emergency Deterrent: requires a *Watchtower* which monitors the bitcoin network and maintains multiple direct communication channels with the stakeholder. If the stakeholder triggers their panic button, a message is sent to the Watchtower which will broadcast all of its Emergency TXs and Unvault Emergency TXs. To further reduce the incentive of a theft attempt the broadcast of any Emergency TX associated with any vault triggers the broadcast of all Emergency TXs associated with all the remaining vaults.

Reviewer 2: Bitcoin

Reviewer 2: Panic button is only one example. One could have proof-of-life automatic emergency, or anything else, some automated.

> Automatic Emergency Deterrent: requires a *Watchtower* which monitors the bitcoin network and maintains multiple direct communication channels with the stakeholder. If the stakeholder triggers their panic button, a message is sent to the Watchtower which will broadcast all of its Emergency TXs and Unvault Emergency TXs. To further reduce the incentive of a theft attempt the broadcast of any Emergency TX associated with any vault triggers the broadcast of all Emergency TXs associated with all the remaining vaults.

Author: > 2. Arguably not necessary, at least for what you describe (it is for sharing the sigs, which all policies you describe here need). A reverse dead man switch doesn't need a message to be sent to the WT. In general i find this kind of policies mildly useless: if you were not cut your communication, why not broadcast the txs yourself?

Do you think it's ever useful to have Watchtowers broadcast Emergency TXs *first*? Do you imagine that the 'panic button' will always be in the Revault App (e.g. no separate button with direct connection to WT)?

Please see [here](https://github.com/revault/practical-revault/blob/master/messages.md#processes) for a more detailed look at the processes in Revault. 

> The flexibility brought by the active defense mechanism allows any unvault or spending policy to be set and enforced by watchtowers. This therefore largely extends the scope of the possibilities compared to on-chain enforcement and is more in-phase with custodians’ needs.

Reviewer: It is also more flexible.

How about:
```suggestion
The flexibility brought by the active defense mechanism allows any unvault or spending policy to be set and enforced by watchtowers. This therefore largely extends the scope of the possibilities compared to on-chain enforcement and is more in-phase with typical business policies (enforcement of a 2FA, of business hours, etc..).
```

## Threat Model

> While Revault has been designed to fix (some of) the threats on existing custody protocols, it introduces new challenges and different assumptions. Revault aims to prevent theft with up to N-1 stakeholders' signing devices being compromised. Here we highlight different attack scenarios, trying to identify weaknesses in the design and finish with a set of security assumptions introduced when using Revault. For more detailed research, check the [Operational Risk Framework](https://github.com/revault/research/blob/master/risk_analysis/paper.pdf).

Reviewer: :+1:

### Emergency Deterrent Risk

While the Emergency TX is the strongest mitigation in the Revault architecture, it introduces a potential risk of blackmail or losses derived from lack of business continuity (e.g. a block-and-short attack). An attacker (internal or external) getting hold of a single Emergency TX can threaten to push it to the Bitcoin network, triggering a push of all Emergency TXs by the watchtowers. The funds are not at risk in this scenario but getting the funds back online would be long and difficult. On the other hand, an important part of the security model (the deterrent)   is enforced by the pre-signed Emergency TXs being accessible.

### Delegation Risk

An "active defence" mechanism is used to restrict the behaviour of fund managers. Contrary to commonly deployed multi-signature wallets, it takes managers two transactions to spend funds. The first step encodes an on-chain time-lock. During this time, only a single honest watchtower is required to cancel given a breach of the unvault policy. It is expected and encouraged to deploy many watchtowers (at least one per stakeholder). 

> To by-pass the spend policy enforcement, an attacker would need to compromise all co-signing servers or to compromise all watchtowers. In the former case, an attacker could alter the Spend TX after the Unvault time-lock expires. In the latter case, a malicious Spend TX would not trigger the broadcast of a Cancel TX.

Reviewer 2: Maybe we can talk about the "tiers" of security here. Unvault policies being first class, cosig being of lesser security due to weaker assumptions (cosigs behaving)

### Denial-of-service Risk

As a consequence of the N-1 resilience model, there are multiple attacks against the practical usage of the funds:

> - signing refusal by a stakeholder

Reviewer 2: (deposit)

- unjustified Cancel TX push
- unjustified Emergency TX push
- taking down a co-signing server

> - taking down the coordinator

Reviewer 2: sure, but I guess the Coord should have redundancy/high availability.

> In addition, if the network becomes highly congested, high feerates would trigger nodes to increase their minimum accepted feerate. Unvault TXs that are prepared with an initial 253 sat/kWU feerate could fail to be propagated sufficiently. Managers would lose access to funds delegated to them until stakeholders can re-sign Unvault TXs with a higher feerate. 

Reviewer 2: or the mempool add fee lowers.

> In addition, if the network becomes highly congested, high feerates would trigger nodes to increase their minimum accepted feerate. Unvault TXs that are prepared with an initial 253 sat/kWU feerate could fail to be propagated sufficiently. Managers would lose access to funds delegated to them until stakeholders can re-sign Unvault TXs with a higher feerate. 

Reviewer: ```suggestion
In addition, if the network becomes highly congested, high feerates would trigger most [0] Bitcoin network nodes' to increase their mempool's minimum accepted feerate. Unvault TXs that are prepared with an initial 24 sat/vb feerate could fail to be relayed should the minimum accepted feerate increase sufficiently [1]. Managers would not be able to use funds delegated to them until the backlog of transactions is processed by the Bitcoin network (mined). 

[0] Those who didn't manually increase the already pretty-high mempool limits.
[1] It's exponential, so that's very likely that in case of the mempool getting above 300MB it quickly won't accept the Unvaults at 24sat/vb
```

It's been more than a year that we don't pre-sign Unvault transaction at 1sat/vbyte..

### Key-tree Risk

The usage of a BIP32 unhardened derivation introduce even more reliance of the signing devices security. Since the chaincodes are exchanged and stored on the wallets (running on untrusted online devices) of each participant, leaking a single child private key would be fatal to a participant's entire key-tree. Therefore the leak of a single child private key of all N stakeholders would be equivalent to the loss of all the funds. Using unhardened derivation implies that there is no partial loss of funds in case of a key leak by every stakeholder.

### Transaction Fee Risk

A fundamental issue arises with pre-signed transactions commiting to the feerate well in advance of the time of broadcast. The feerate is unpredictable and a transaction with a low feerate may not be included in a block in time. 

> The Unvault TXs have an additional output that allows dynamically allocating fees through child-pays-for-parent (CPFP). 

Reviewer: Maybe emphase it's not security critical, "only" for usability

> The Cancel, Emergency and Unvault Emergency TXs aren't compatible with CPFP-carve out (FIXME: Why?), nor with ANYONECANPAY|ALL signatures (because this enables pinning attacks). The safest solution is to prepare multiple variants of the same transactions with a broad distribution of feerates, from typical to extreme and with many in-between. 

Reviewer 2: because carve-out is for 2 party.
I don't like to call it "pinning", to me it's a different class of attack. I'll let @darosior find a proper name.
"many in-between" isn't necessary, granularity is a bonus more than a need.

Reviewer: We (people working on this in the Bitcoin space) have always refered to it as pinning, i think it's fine. But i have a comment on this line as well which asks to remove this anyways :)

> The Cancel, Emergency and Unvault Emergency TXs aren't compatible with CPFP-carve out (FIXME: Why?), nor with ANYONECANPAY|ALL signatures (because this enables pinning attacks). The safest solution is to prepare multiple variants of the same transactions with a broad distribution of feerates, from typical to extreme and with many in-between. 

Reviewer: You can even just put a note that it's not detailed as it's going to be removed soon due to pinning concerns.

> ### Underlying Assumptions

Reviewer: I think we should remove that. It needs a more thorough analysis, i think.

#### Bitcoin

> - assume users can propagate revocation transactions to miners.

Reviewer 2: if we did, we wouldn't have the fee discussions all the time :)

- assume miners will mine the transaction paying the most feerate, regardless of any other (meta)data.

> - assume a significant hashrate to prevent reorgs of depth higher than the unvault time-lock length.

Reviewer 2: Yes.

Reviewer 2: Altho it's more complex than that. We don't want "natural reorgs" to be long, but they still statistically can happen. And then there is the game theory as in: are the miners going to mine the next block instead of rewriting the chain?
It's not only a simple hashrate discussion.

#### Infrastructure

- assume a proper ceremony (FIXME: link) as the root of trust for cryptographic protocols used.

> - assume a secure signing device allowing users to check the validity of the data being signed.

Reviewer 2: meh, more like enforcing descriptors, securing keys, and being tamper proof/evident.

> - assume at least one honest cosigning server AND at least one honest online watchtower per stakeholder.

Reviewer 2: Not really, 1 honest WT and 1 honest cosig in total is enough.
Also, does not apply to unvault policies, in which no cosig is necessary.

- assume the N stakeholders’ keys to not all be compromised.

> - assume sufficient fees for Revault TXs such that operations aren't stalled. 

Reviewer 2: is this an assumption or part of the protocol?

The stakeholders exchange signatures for the Cancel, Emergency and Unvault Emergency TXs:

- The Cancel TX would spend the Unvault TX to pay back to the *[deposit output descriptor](transactions.md#out)*.
- The Emergency TX would spend a deposit and pay to the *[emergency output descriptor](transactions.md#emergency_txs)*.
- The Unvault Emergency TX would spend the Unvault TX to pay to the *[emergency output descriptor](transactions.md#emergency_txs)*.

This way, funds are protected _before_ delegation to fund managers. Eventually, the Unvault TX may be signed (to restrictively delegate the UTXO to managers). It spends from the *[deposit output descriptor](transactions.md#out)* and pays to the *[unvault output descriptor](transactions.md#out-1)*, and is accessible with either:

- the managers' K-of-M multi-signature + time-lock of X blocks + J cosigning servers (where J = 0 if deploying without co-signers or N of deploying with co-signers), or
- the stakeholders' N-of-N multi-signature (with no delay).

> Managers broadcast the relevant Unvault TX, and can attempt to use the Spend TX to pay external wallets, but are forced to wait for the Unvault time-lock to expire before their transaction finalizes. During that delay period, a Cancel or Unvault Emergency TX can be broadcasted which will revoke the spend attempt.

Reviewer: I'm sorry for nit'ing again.. I really did not want to but i find this too misleading.

> Managers broadcast the relevant Unvault TX, and can attempt to use the Spend TX to pay external wallets,, but are forced to wait for the Unvault time-lock to expire before their transaction finalizes.

Is it me or this seems to imply the managers broadcast the Spend right after broadcasting the Unvault?
Also, if we keep it like that i'd at least say "to expire before the Spend transaction becomes valid".

First, there are base requirements common to all Revault deployments. Each stakeholder and manager controls a wallet and a signing device. An untrusted Coordinator server is used to route messages and thus enable wallets to update and synchronise their state.  

- **Revault Wallet**: an open-source software running with a bitcoind back-end used to keep track of the users' co-owned coins, transaction signatures, transaction history and to connect with a signing device and infrastructure servers.

> - **Signing Device**: a hardware module used to store private keys and wallet descriptors and to handle signing operations. Given information of the UTXO locked to the *[deposit output descriptor](transactions.md#out)*, this device can construct the set of Revault TXs itself based on its configuration data. This alleviates the need for human-verification of the TXs when signing.

Reviewer: :+1:

> - **Coordinator server**: An always-online server used for efficient message routing among participants and other servers. Its primary use is for exchanging signatures of pre-signed transactions and Spend TXs.

Reviewer: nit: If you retouch, can you make this:
> An always-online *untrusted* server

With these in place, stakeholders are able to delegate bitcoin to fund managers and manually broadcast Cancel, Emergency and Unvault-Emergency TXs. Fund managers will also be able to manually broadcast Cancel TXs to recover from mistakes. 

Next, users who want additional features will require additional infrastructure components. Each stakeholder may control one or more watchtowers and/ or a co-signing server.

> - **Watchtower**: An always-online server that monitors the Bitcoin network and chain, and depending on its responsibilities (which could include; automated unvault policy enforcement, automated spend policy enforcement, emergency deterrent) it will respond to relevant events.

Reviewer: Perfect

Reviewer: The use of `;` in the english language i've always found interesting

Author: ah I actually used it wrong. It should be `:`

- unjustified Cancel TX push
- unjustified Emergency TX push
- taking down a co-signing server
- taking down (all) the coordinator(s)
- pinning Unvault TXs

In addition, if the network becomes highly congested, high feerates would trigger most Bitcoin network nodes' to increase their mempool's minimum accepted feerate (those who didn't manually increase the mempool limits). Unvault TXs that are prepared with an initial 24 sat/vb feerate could fail to be relayed if the minimum accepted feerate increases sufficiently. Managers would not be able to use funds delegated to them until the backlog of transactions is processed by the Bitcoin network (mined).

### Key-tree Risk

> The usage of a BIP32 unhardened derivation introduce even more reliance on the signing devices' security. Since the chaincodes are exchanged and stored on the wallets (running on untrusted online devices) of each participant, leaking a single child private key would be fatal to a participant's entire key-tree. Therefore the leak of a single child private key of all N stakeholders would be equivalent to the loss of all the funds. Using unhardened derivation implies that there is no partial loss of funds in case of a key leak by every stakeholder.

Reviewer: nit: introduce*s*?

### Transaction Fee Risk

A fundamental issue arises with pre-signed transactions commiting to the feerate well in advance of the time of broadcast. The required feerate is unpredictable and a transaction with a low feerate may not be included in a block in time.

[Pinning attacks](https://bitcoinops.org/en/topics/transaction-pinning/) are possible when using either RBF or CPFP to dynamically allocate fees to transactions. These attacks exploit vulnerabilities that can cause transactions to become "stuck" in the mempool for long times. For protocols like Revault which require timely transaction confirmation, these attacks can be used to defeat theft-mitigation features (e.g. pinning a Cancel TX for longer than the Unvault TX's time-lock, and stealing funds).

> The Unvault TXs have an additional output that allows dynamically allocating fees through child-pays-for-parent (CPFP). This feature helps managers optimise withdrawal times. The possibility of pinning Unvault TXs doesn't impact the theft-mitigation features of Revault but can be used as a denial-of-service attack by comrpomised managers. This is acceptable since managers already have multiple ways to launch denial-of-service attacks. For the Cancel, Emergency and Unvault Emergency TXs it is critical to avoid transaction pinning attacks. Unfortunately, these transactions aren't compatible with Bitcoin's [CPFP carve-out](https://github.com/bitcoin/bitcoin/pull/15681) rule which is used to mitigate pinning attacks in Lightning.

Reviewer: Instead of the specific case of the carve-out you could link to https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2021-November/019614.html, saying we did not find a non-pinable way of having dynamic fee bumping on today's Bitcoin.

> Emergency and Unvault Emergency TXs are simply pre-signed with exorbitant fees as they are intended as a deterrent, and not a regular expense. Since Cancel TXs may (or may not) occur frequently, it is important to minimise their feerates while still ensuring timely confirmation. The best solution currently is to prepare multiple variants of the same Cancel TX with a broad distribution of feerates, from typical to extreme and with many in-between. At the time of broadcast, the one with the appropriate fee should be selected.

Reviewer: Meh, it's not what the specs are saying. Let's keep that to when we modify the spec to not use ACP anymore. Maybe even the previous paragraph.

Author: Ok I removed it for now
